File: apps/felix_web/assets/js/phasertron/utils/lua/GOAP.lua
GoapUtils = {}

function GoapUtils.statesEqual(a, b)
    for key, value in pairs(a) do
        if b[key] ~= value then
            return false
        end
    end
    for key, value in pairs(b) do
        if a[key] ~= value then
            return false
        end
    end
    return true
end

function GoapUtils.calculateEffectsDistance(action, state)
    local dist = 0
    for effectKey, fxVal in pairs(action.effects) do
        if state[effectKey] == nil then
            dist = dist + math.abs(fxVal)
        else
            if fxVal < state[effectKey] then
                dist = dist + math.abs(fxVal - state[effectKey])
            end
        end
    end

    for stateKey, stateValue in pairs(state) do
        if action.effects[stateKey] == nil then
            dist = dist + math.abs(stateValue)
        end
    end
    return dist
end

function GoapUtils.actionCanContribute(action, state)
    for key, stateValue in pairs(state) do
        local fxValue = action.effects[key] or 0
        local sValue = stateValue or 0
        if fxValue > 0 or fxValue >= sValue then
            return true
        end
        if fxValue < 0 and sValue <= fxValue then
            return true
        end
    end
    return false
end

function GoapUtils.getBadnessScore(thread)
    local score = thread.totalCost
    local numActions = #thread.plan
    score = score + numActions
    for i = 1, numActions - 1 do
        if thread.plan[i].name == thread.plan[i + 1].name then
            score = score - 10
        end
    end
    return score
end

GoapPlanner = {}
function GoapPlanner:new()
    newObj = {
        queue = DoublyLinkedList:new()
    }
    self.__index = self
    return setmetatable(newObj, self)
end

function GoapPlanner:plan(actions, startState, goalState)
    print('Planning...')

    -- Sort actions by cost
    table.sort(actions, function(a, b)
        return a.cost < b.cost
    end)

    self.queue = DoublyLinkedList:new() -- Re-initialize the queue for each planning attempt
    self.queue:insertFirst({
        plan = {},
        state = goalState,
        totalCost = 0
    })

    local considerCount = 0
    while not self.queue:isEmpty() do
        local current = self.queue:removeFirst()
        considerCount = considerCount + 1

        if considerCount > 1000 then
            print('Considered ' .. considerCount .. ' plans, giving up.')
            return nil
        end

        for _, action in ipairs(actions) do
            if GoapUtils.actionCanContribute(action, current.state) then
                local newState = {}
                for k, v in pairs(current.state) do
                    newState[k] = v
                end

                -- Apply effects
                for key, val in pairs(action.effects) do
                    newState[key] = (newState[key] or 0) - val
                    if newState[key] <= 0 then
                        newState[key] = nil
                    end
                end

                -- Merge preconditions into the new state
                for key, val in pairs(action.preconditions) do
                    newState[key] = math.max(newState[key] or 0, val)
                    if newState[key] <= 0 then
                        newState[key] = nil
                    end
                end

                local newPlan = {table.unpack(current.plan)}

                table.insert(newPlan, action)

                local dist = GoapUtils.calculateEffectsDistance(action, current.state)
                local nextThread = {
                    plan = newPlan,
                    state = newState,
                    totalCost = current.totalCost + dist + (action.cost * 10)
                }

                if GoapUtils.statesEqual(newState, startState) then
                    -- Reverse the plan to match the original action sequence
                    local finalPlan = {}

                    for i = #nextThread.plan, 1, -1 do
                        table.insert(finalPlan, nextThread.plan[i])
                    end
                    print('Found plan after ' .. considerCount .. ' considerations!')
                    return finalPlan
                end

                -- heuristic
                local head = self.queue:getHead()
                if not head then
                    self.queue:insertLast(nextThread)
                else
                    local headScore = GoapUtils.getBadnessScore(head)
                    local nextScore = GoapUtils.getBadnessScore(nextThread)
                    if headScore <= nextScore then
                        self.queue:insertLast(nextThread)
                    else
                        self.queue:insertFirst(nextThread)
                    end
                end
            end
        end
    end

    print('No plan found after ' .. considerCount .. ' considerations.')
    return nil
end

-- Utility function for reversing tables
function reverseTable(tbl)
    local size = #tbl
    local newTbl = {}
    for i, v in ipairs(tbl) do
        newTbl[size - i + 1] = v
    end
    return newTbl
end

-- Example actions setup
-- actions = {{
--     preconditions = {},
--     effects = {
--         hasCover = 1
--     },
--     cost = 1,
--     name = 'findCover'
-- }, {
--     preconditions = {
--         hasCover = 1
--     },
--     effects = {
--         isHidden = 1
--     },
--     cost = 1,
--     name = 'hide'
-- }, {
--     preconditions = {},
--     effects = {
--         hasShots = 1
--     },
--     cost = 3,
--     name = 'reload'
-- }, {
--     preconditions = {
--         hasShots = 1
--     },
--     effects = {
--         hasVisual = 1
--     },
--     cost = 2,
--     name = 'searchForOpponent'
-- }, {
--     preconditions = {
--         hasShots = 1,
--         hasVisual = 1
--     },
--     effects = {
--         opponentHit = 1,
--         hasShots = -1
--     },
--     cost = 2,
--     name = 'shoot'
-- }, {
--     preconditions = {
--         opponentHit = 1
--     },
--     effects = {
--         canTaunt = 1
--     },
--     cost = 1,
--     name = 'taunt'
-- }, {
--     preconditions = {},
--     effects = {
--         atStartPoint = 1
--     },
--     cost = 5,
--     name = 'returnToStart'
-- }, {
--     preconditions = {
--         hasCover = 1,
--         isHidden = 0
--     },
--     effects = {
--         isHidden = 1
--     },
--     cost = 2,
--     name = 'duckAndHide'
-- }, {
--     preconditions = {
--         hasShots = 0
--     },
--     effects = {
--         seekAmmo = 1
--     },
--     cost = 3,
--     name = 'seekAmmo'
-- }, {
--     preconditions = {
--         seekAmmo = 1
--     },
--     effects = {
--         hasShots = 1
--     },
--     cost = 2,
--     name = 'pickUpAmmo'
-- }, {
--     preconditions = {
--         isHidden = 1,
--         hasVisual = 0
--     },
--     effects = {
--         hasVisual = 1
--     },
--     cost = 3,
--     name = 'peekForOpponent'
-- }}

File: apps/felix_web/assets/js/phasertron/utils/ai/goap/GOAP.ts
import { GoapAction, GoapState } from './GoapTypes';
import GoapValidator from './GoapActionValidator';
import DoublyLinkedList from '../../data-structures/DoublyLinkedList';
import { BehaviorTree } from '../behavior-trees/BehaviorTree';
import { Selector } from '../behavior-trees/base/Selector';
import GoapRunner from './GoapRunner';
import { AlwaysSucceed } from '../behavior-trees/decorators/AlwaysSucceed';

class GoapUtils {
  public static statesEqual = (a: GoapState, b: GoapState): boolean => {
    for (const key in a) {
      if (a[key] !== b[key]) {
        return false;
      }
    }
    for (const key in b) {
      if (a[key] !== b[key]) {
        return false;
      }
    }
    return true;
  };

  public static calculateEffectsDistance = (
    action: GoapAction,
    state: GoapState
  ): number => {
    let dist = 0;

    // Increase score for extraneous effects not needed by the state
    for (const effectKey in action.effects) {
      const fxVal = action.effects[effectKey];
      if (!(effectKey in state)) {
        dist += Math.abs(fxVal);
      } else {
        if (fxVal < state[effectKey]) {
          dist += Math.abs(fxVal - state[effectKey]);
        }
      }
    }

    // Increase score for missing effects that are needed by the state but not provided by the action
    for (const stateKey in state) {
      if (!(stateKey in action.effects)) {
        dist += Math.abs(state[stateKey]);
      }
    }

    return dist;
  };

  public static actionCanContribute = (
    action: GoapAction,
    state: GoapState
  ): boolean => {
    // an action can contribute to a state if the action's `effects` can contribute to the state
    // this means a positive effect can only contribute if the state has >= the effect's value already
    // and a negative effect can only contribute if the state has <= the effect's value
    for (const key in state) {
      const fxValue = action.effects[key] ?? 0;
      const stateValue = state[key] ?? 0;

      if (fxValue > 0 || fxValue >= stateValue) {
        return true;
      }
      if (fxValue < 0 && stateValue <= fxValue) {
        return true;
      }
    }
    return false;
  };

  public static getBadnessScore = (thread: GoapThread): number => {
    let score = thread.totalCost;
    const numActions = thread.plan.length;

    // prefer plans that are shorter
    score += numActions;

    // prefer plans that perform multiple actions sequentially versus ping-ponging between others
    // i.e. [goFish, goFish, sellFish] is better than [goFish, sellFish, goFish, sellFish]
    for (let i = 0; i < numActions - 1; i++) {
      if (thread.plan[i].name === thread.plan[i + 1].name) {
        score -= 10;
      }
    }

    return score;
  };
}

type GoapThread = {
  plan: GoapAction[];
  state: GoapState;
  totalCost: number;
};

export class GoapPlanner {
  public plan = (
    actions: GoapAction[],
    startState: GoapState,
    goalState: GoapState
  ): GoapAction[] => {
    const queue = new DoublyLinkedList<GoapThread>();
    queue.insertFirst({
      plan: [],
      state: goalState,
      totalCost: 0,
    });

    console.log('startState', startState);
    console.log('goalState', goalState);

    let considerCount = 0;
    let current;
    while ((current = queue.removeFirst())) {
      considerCount++;

      for (const action of actions) {
        // if current state has all of the *effects* for this action, then we can add it to the plan
        if (GoapUtils.actionCanContribute(action, current.state)) {
          const newState = { ...current.state };

          // then we'll add its *preconditions* to the current state and add it to the queue
          for (const key in action.effects) {
            newState[key] ??= 0;
            newState[key] -= action.effects[key];
            newState[key] = Math.max(0, newState[key]);
            if (newState[key] === 0) {
              delete newState[key];
            }
          }
          for (const key in action.preconditions) {
            newState[key] ??= 0;
            newState[key] = Math.max(newState[key], action.preconditions[key]);
            newState[key] = Math.max(0, newState[key]);
            if (newState[key] === 0) {
              delete newState[key];
            }
          }

          const dist = GoapUtils.calculateEffectsDistance(
            action,
            current.state
          );

          const nextThread = {
            plan: [...current.plan, action],
            state: newState,
            totalCost: current.totalCost + dist * 10 + action.cost * 100,
          };

          if (GoapUtils.statesEqual(newState, startState)) {
            console.log('Found plan after ' + considerCount + ' threads!');
            return nextThread.plan.reverse();
          }

          // heuristic
          const head = queue.head?.data;
          if (!head) {
            queue.insertLast(nextThread);
          } else {
            const headScore = GoapUtils.getBadnessScore(head);
            const nextScore = GoapUtils.getBadnessScore(nextThread);

            if (headScore <= nextScore) {
              queue.insertLast(nextThread);
            } else {
              queue.insertFirst(nextThread);
            }
          }
        }
      }
    }

    return [];
  };
}

let actions: GoapAction[] = [
  // collect sticks
  {
    preconditions: {},
    effects: { hasSticks: 1 },
    cost: 1,
    name: 'collectSticks',
    behaviorTree: new BehaviorTree(new AlwaysSucceed(new Selector())),
  },
  // collect rocks
  {
    preconditions: {},
    effects: { hasRocks: 1 },
    cost: 1,
    name: 'collectRocks',
    behaviorTree: new BehaviorTree(new AlwaysSucceed(new Selector())),
  },
  // make an axe from sticks + rocks
  {
    preconditions: { hasSticks: 1, hasRocks: 1 },
    effects: { hasSticks: -1, hasRocks: -1, hasAxe: 1 },
    cost: 1,
    name: 'makeAxe',
    behaviorTree: new BehaviorTree(new AlwaysSucceed(new Selector())),
  },
  // chop wood
  {
    preconditions: { hasAxe: 1 },
    effects: { hasWood: 1 },
    cost: 1,
    name: 'chopWood',
    behaviorTree: new BehaviorTree(new AlwaysSucceed(new Selector())),
  },
];
actions = actions.sort((a, b) => a.cost - b.cost);

const startState = {};
const goalState = { hasWood: 1 };
const planner = new GoapPlanner();
const plan = planner.plan(actions, startState, goalState);

// console.time('everything');

// console.time('create validator');
// const validator = new GoapValidator(actions);
// console.timeEnd('create validator');

// console.time('validateActions');
// validator.validateActions(true);
// console.timeEnd('validateActions');

// console.time('validateStates');
// validator.validateState(startState, true);
// validator.validateState(goalState, true);
// console.timeEnd('validateStates');

// console.time('planning');
// console.timeEnd('planning');
// console.timeEnd('everything');
// console.log('plan', plan);

// // "replay" the action on a new startstate to see if we actually meet the end goal
// const newState: { [id: string]: number } = { ...startState };
// console.log('\n\n\nInitial state', newState, '\tDesired state', goalState);
// for (const action of plan) {
//   for (const key in action.effects) {
//     newState[key] ??= 0;
//     newState[key] += action.effects[key];
//     if (newState[key] === 0) {
//       delete newState[key];
//     }
//   }
//   console.log('Next action: ' + action.name, newState);
// }
// console.log('Final result', newState);

const runner = new GoapRunner(planner, plan);
// console.log('running plan..');
// setInterval(() => {
//   runner.update();
// }, 1000);
